commit 3c3c01f9e5da55f8843e5d111a430bc75af4c600
Author: Ola x Nilsson <olani@axis.com>
Date:   Fri Mar 17 08:44:44 2023 +0100

    gnu: Use _TIME_BITS and _FILE_OFFSET_BITS=64 versions of glibc symbols
    
    Set the link names of relevant symbols to use be the same as when a C
    program is built against GNU libc with -D_TIME_BITS=64 and
    -D_FILE_OFFSET_BITS=64.

--- a/src/unix/linux_like/linux/gnu/b32/mod.rs
+++ b/src/unix/linux_like/linux/gnu/b32/mod.rs
@@ -340,2 +340,4 @@
     ) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "__ntp_gettime64")]
+    pub fn ntp_gettime(buf: *mut ntptimeval) -> ::c_int;
 }
--- a/src/unix/linux_like/linux/gnu/b64/mod.rs
+++ b/src/unix/linux_like/linux/gnu/b64/mod.rs
@@ -99,2 +99,7 @@
 
+extern "C" {
+    #[link_name = "ntp_gettimex"]
+    pub fn ntp_gettime(buf: *mut ntptimeval) -> ::c_int;
+}
+
 cfg_if! {
--- a/src/unix/linux_like/linux/gnu/mod.rs
+++ b/src/unix/linux_like/linux/gnu/mod.rs
@@ -1300,2 +1300,3 @@
     );
+    #[cfg_attr(gnu_time64_abi, link_name = "__sendmmsg64")]
     pub fn sendmmsg(
@@ -1306,2 +1307,3 @@
     ) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "__recvmmsg64")]
     pub fn recvmmsg(
@@ -1316,4 +1318,7 @@
     pub fn setrlimit64(resource: ::__rlimit_resource_t, rlim: *const ::rlimit64) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "getrlimit64")]
     pub fn getrlimit(resource: ::__rlimit_resource_t, rlim: *mut ::rlimit) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "setrlimit64")]
     pub fn setrlimit(resource: ::__rlimit_resource_t, rlim: *const ::rlimit) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "prlimit64")]
     pub fn prlimit(
@@ -1340,2 +1345,3 @@
     pub fn mallopt(param: ::c_int, value: ::c_int) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "__gettimeofday64")]
     pub fn gettimeofday(tp: *mut ::timeval, tz: *mut ::timezone) -> ::c_int;
@@ -1352,6 +1358,7 @@
 
+    #[cfg_attr(gnu_time64_abi, link_name = "___adjtimex64")]
     pub fn adjtimex(buf: *mut timex) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "___adjtimex64")]
     pub fn ntp_adjtime(buf: *mut timex) -> ::c_int;
-    #[link_name = "ntp_gettimex"]
-    pub fn ntp_gettime(buf: *mut ntptimeval) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "__clock_adjtime64")]
     pub fn clock_adjtime(clk_id: ::clockid_t, buf: *mut ::timex) -> ::c_int;
@@ -1365,2 +1372,3 @@
     ) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "preadv64v2")]
     pub fn preadv2(
@@ -1372,2 +1380,3 @@
     ) -> ::ssize_t;
+    #[cfg_attr(gnu_time64_abi, link_name = "pwritev64v2")]
     pub fn pwritev2(
@@ -1408,2 +1417,3 @@
     pub fn backtrace(buf: *mut *mut ::c_void, sz: ::c_int) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "__glob64_time64")]
     pub fn glob64(
@@ -1414,2 +1424,3 @@
     ) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "__globfree64_time64")]
     pub fn globfree64(pglob: *mut glob64_t);
@@ -1480,2 +1491,3 @@
     pub fn asctime_r(tm: *const ::tm, buf: *mut ::c_char) -> *mut ::c_char;
+    #[cfg_attr(gnu_time64_abi, link_name = "__ctime64_r")]
     pub fn ctime_r(timep: *const time_t, buf: *mut ::c_char) -> *mut ::c_char;
--- a/src/unix/linux_like/linux/mod.rs
+++ b/src/unix/linux_like/linux/mod.rs
@@ -4852,7 +4852,12 @@
         extern "C" {
+            #[cfg_attr(gnu_time64_abi, link_name = "aio_read64")]
             pub fn aio_read(aiocbp: *mut aiocb) -> ::c_int;
+            #[cfg_attr(gnu_time64_abi, link_name = "aio_write64")]
             pub fn aio_write(aiocbp: *mut aiocb) -> ::c_int;
             pub fn aio_fsync(op: ::c_int, aiocbp: *mut aiocb) -> ::c_int;
+            #[cfg_attr(gnu_time64_abi, link_name = "aio_error64")]
             pub fn aio_error(aiocbp: *const aiocb) -> ::c_int;
+            #[cfg_attr(gnu_time64_abi, link_name = "aio_return64")]
             pub fn aio_return(aiocbp: *mut aiocb) -> ::ssize_t;
+            #[cfg_attr(gnu_time64_abi, link_name = "__aio_suspend_time64")]
             pub fn aio_suspend(
@@ -4862,3 +4867,5 @@
             ) -> ::c_int;
+            #[cfg_attr(gnu_time64_abi, link_name = "aio_cancel64")]
             pub fn aio_cancel(fd: ::c_int, aiocbp: *mut aiocb) -> ::c_int;
+            #[cfg_attr(gnu_time64_abi, link_name = "lio_listio64")]
             pub fn lio_listio(
@@ -4876,2 +4883,3 @@
         extern "C" {
+            #[cfg_attr(gnu_time64_abi, link_name = "pwritev64")]
             pub fn pwritev(
@@ -4882,2 +4890,3 @@
             ) -> ::ssize_t;
+            #[cfg_attr(gnu_time64_abi, link_name = "preadv64")]
             pub fn preadv(
@@ -4917,2 +4926,3 @@
             ) -> isize;
+            #[cfg_attr(gnu_time64_abi, link_name = "__futimes64")]
             pub fn futimes(
@@ -4949,2 +4959,3 @@
             ) -> ::ssize_t;
+            #[cfg_attr(gnu_time64_abi, link_name = "__mq_timedreceive_time64")]
             pub fn mq_timedreceive(
@@ -4962,2 +4973,3 @@
             ) -> ::c_int;
+            #[cfg_attr(gnu_time64_abi, link_name = "__mq_timedsend_time64")]
             pub fn mq_timedsend(
@@ -5012,2 +5024,3 @@
 
+    #[cfg_attr(gnu_time64_abi, link_name = "__lutimes64")]
     pub fn lutimes(file: *const ::c_char, times: *const ::timeval) -> ::c_int;
@@ -5033,2 +5046,3 @@
     pub fn shmdt(shmaddr: *const ::c_void) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "__shmctl64")]
     pub fn shmctl(shmid: ::c_int, cmd: ::c_int, buf: *mut ::shmid_ds) -> ::c_int;
@@ -5037,3 +5051,5 @@
     pub fn semop(semid: ::c_int, sops: *mut ::sembuf, nsops: ::size_t) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "__semctl64")]
     pub fn semctl(semid: ::c_int, semnum: ::c_int, cmd: ::c_int, ...) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "__msgctl64")]
     pub fn msgctl(msqid: ::c_int, cmd: ::c_int, buf: *mut msqid_ds) -> ::c_int;
@@ -5057,3 +5073,5 @@
 
+    #[cfg_attr(gnu_time64_abi, link_name = "fallocate64")]
     pub fn fallocate(fd: ::c_int, mode: ::c_int, offset: ::off_t, len: ::off_t) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "posix_fallocate64")]
     pub fn posix_fallocate(fd: ::c_int, offset: ::off_t, len: ::off_t) -> ::c_int;
@@ -5107,3 +5125,5 @@
     pub fn timerfd_create(clockid: ::clockid_t, flags: ::c_int) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "__timerfd_gettime64")]
     pub fn timerfd_gettime(fd: ::c_int, curr_value: *mut itimerspec) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "__timerfd_settime64")]
     pub fn timerfd_settime(
@@ -5128,4 +5148,7 @@
     pub fn dup3(oldfd: ::c_int, newfd: ::c_int, flags: ::c_int) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "mkostemp64")]
     pub fn mkostemp(template: *mut ::c_char, flags: ::c_int) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "mkostemps64")]
     pub fn mkostemps(template: *mut ::c_char, suffixlen: ::c_int, flags: ::c_int) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "__sigtimedwait64")]
     pub fn sigtimedwait(
@@ -5176,2 +5199,3 @@
 
+    #[cfg_attr(gnu_time64_abi, link_name = "__glob64_time64")]
     pub fn glob(
@@ -5182,2 +5206,3 @@
     ) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "__globfree64_time64")]
     pub fn globfree(pglob: *mut ::glob_t);
@@ -5207,2 +5232,3 @@
     ) -> ::ssize_t;
+    #[cfg_attr(gnu_time64_abi, link_name = "mkstemps64")]
     pub fn mkstemps(template: *mut ::c_char, suffixlen: ::c_int) -> ::c_int;
@@ -5243,2 +5269,3 @@
     pub fn tee(fd_in: ::c_int, fd_out: ::c_int, len: ::size_t, flags: ::c_uint) -> ::ssize_t;
+    #[cfg_attr(gnu_time64_abi, link_name = "__settimeofday64")]
     pub fn settimeofday(tv: *const ::timeval, tz: *const ::timezone) -> ::c_int;
@@ -5256,3 +5283,5 @@
 
+    #[cfg_attr(gnu_time64_abi, link_name = "__sched_rr_get_interval64")]
     pub fn sched_rr_get_interval(pid: ::pid_t, tp: *mut ::timespec) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "__sem_timedwait64")]
     pub fn sem_timedwait(sem: *mut sem_t, abstime: *const ::timespec) -> ::c_int;
@@ -5276,4 +5305,6 @@
     pub fn personality(persona: ::c_ulong) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "__prctl_time64")]
     pub fn prctl(option: ::c_int, ...) -> ::c_int;
     pub fn sched_getparam(pid: ::pid_t, param: *mut ::sched_param) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "__ppoll64")]
     pub fn ppoll(
@@ -5293,2 +5324,3 @@
 
+    #[cfg_attr(gnu_time64_abi, link_name = "__pthread_mutex_timedlock64")]
     pub fn pthread_mutex_timedlock(
@@ -5327,2 +5359,10 @@
     pub fn sched_getscheduler(pid: ::pid_t) -> ::c_int;
+    #[cfg_attr(
+        all(
+            target_env = "gnu",
+            target_pointer_width = "32",
+            not(target_arch = "x86_64")
+        ),
+        link_name = "__clock_nanosleep_time64"
+    )]
     pub fn clock_nanosleep(
@@ -5378,2 +5418,3 @@
     ) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "sendfile64")]
     pub fn sendfile(
@@ -5596,3 +5637,5 @@
     pub fn timer_getoverrun(timerid: ::timer_t) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "__timer_gettime64")]
     pub fn timer_gettime(timerid: ::timer_t, curr_value: *mut ::itimerspec) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "__timer_settime64")]
     pub fn timer_settime(
@@ -5638,2 +5681,3 @@
 
+    #[cfg_attr(gnu_time64_abi, link_name = "__ioctl_time64")]
     pub fn ioctl(fd: ::c_int, request: ::Ioctl, ...) -> ::c_int;
--- a/src/unix/linux_like/mod.rs
+++ b/src/unix/linux_like/mod.rs
@@ -1677,4 +1677,7 @@
 
+    #[cfg_attr(gnu_time64_abi, link_name = "__clock_getres64")]
     pub fn clock_getres(clk_id: ::clockid_t, tp: *mut ::timespec) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "__clock_gettime64")]
     pub fn clock_gettime(clk_id: ::clockid_t, tp: *mut ::timespec) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "__clock_settime64")]
     pub fn clock_settime(clk_id: ::clockid_t, tp: *const ::timespec) -> ::c_int;
@@ -1693,7 +1696,12 @@
     pub fn pipe2(fds: *mut ::c_int, flags: ::c_int) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "statfs64")]
     pub fn statfs(path: *const ::c_char, buf: *mut statfs) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "fstatfs64")]
     pub fn fstatfs(fd: ::c_int, buf: *mut statfs) -> ::c_int;
     pub fn memrchr(cx: *const ::c_void, c: ::c_int, n: ::size_t) -> *mut ::c_void;
+    #[cfg_attr(gnu_time64_abi, link_name = "posix_fadvise64")]
     pub fn posix_fadvise(fd: ::c_int, offset: ::off_t, len: ::off_t, advise: ::c_int) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "__futimens64")]
     pub fn futimens(fd: ::c_int, times: *const ::timespec) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "__utimensat64")]
     pub fn utimensat(
@@ -1748,2 +1756,3 @@
     pub fn setresuid(ruid: ::uid_t, euid: ::uid_t, suid: ::uid_t) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "__wait4_time64")]
     pub fn wait4(
@@ -1772,3 +1781,5 @@
 
+    #[cfg_attr(gnu_time64_abi, link_name = "__sendmsg64")]
     pub fn sendmsg(fd: ::c_int, msg: *const ::msghdr, flags: ::c_int) -> ::ssize_t;
+    #[cfg_attr(gnu_time64_abi, link_name = "__recvmsg64")]
     pub fn recvmsg(fd: ::c_int, msg: *mut ::msghdr, flags: ::c_int) -> ::ssize_t;
@@ -1806,3 +1817,5 @@
             pub fn creat64(path: *const c_char, mode: mode_t) -> ::c_int;
+            #[cfg_attr(gnu_time64_abi, link_name = "__fstat64_time64")]
             pub fn fstat64(fildes: ::c_int, buf: *mut stat64) -> ::c_int;
+            #[cfg_attr(gnu_time64_abi, link_name = "__fstatat64_time64")]
             pub fn fstatat64(
@@ -1815,2 +1828,3 @@
             pub fn lseek64(fd: ::c_int, offset: off64_t, whence: ::c_int) -> off64_t;
+            #[cfg_attr(gnu_time64_abi, link_name = "__lstat64_time64")]
             pub fn lstat64(path: *const c_char, buf: *mut stat64) -> ::c_int;
@@ -1850,2 +1864,3 @@
             ) -> ::c_int;
+            #[cfg_attr(gnu_time64_abi, link_name = "__stat64_time64")]
             pub fn stat64(path: *const c_char, buf: *mut stat64) -> ::c_int;
--- a/src/unix/mod.rs
+++ b/src/unix/mod.rs
@@ -457,2 +457,3 @@
     )]
+    #[cfg_attr(gnu_time64_abi, link_name = "fopen64")]
     pub fn fopen(filename: *const c_char, mode: *const c_char) -> *mut FILE;
@@ -462,2 +463,3 @@
     )]
+    #[cfg_attr(gnu_time64_abi, link_name = "freopen64")]
     pub fn freopen(filename: *const c_char, mode: *const c_char, file: *mut FILE) -> *mut FILE;
@@ -468,2 +470,3 @@
     pub fn rename(oldname: *const c_char, newname: *const c_char) -> c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "tmpfile64")]
     pub fn tmpfile() -> *mut FILE;
@@ -493,4 +496,6 @@
     #[cfg_attr(target_os = "netbsd", link_name = "__fgetpos50")]
+    #[cfg_attr(gnu_time64_abi, link_name = "fgetpos64")]
     pub fn fgetpos(stream: *mut FILE, ptr: *mut fpos_t) -> c_int;
     #[cfg_attr(target_os = "netbsd", link_name = "__fsetpos50")]
+    #[cfg_attr(gnu_time64_abi, link_name = "fsetpos64")]
     pub fn fsetpos(stream: *mut FILE, ptr: *const fpos_t) -> c_int;
@@ -671,2 +676,3 @@
     #[cfg_attr(target_os = "espidf", link_name = "lwip_setsockopt")]
+    #[cfg_attr(gnu_time64_abi, link_name = "__setsockopt64")]
     pub fn setsockopt(
@@ -737,2 +743,3 @@
     )]
+    #[cfg_attr(gnu_time64_abi, link_name = "__fstat64_time64")]
     pub fn fstat(fildes: ::c_int, buf: *mut stat) -> ::c_int;
@@ -750,2 +757,3 @@
     )]
+    #[cfg_attr(gnu_time64_abi, link_name = "__stat64_time64")]
     pub fn stat(path: *const c_char, buf: *mut stat) -> ::c_int;
@@ -764,2 +772,3 @@
     )]
+    #[cfg_attr(gnu_time64_abi, link_name = "open64")]
     pub fn open(path: *const c_char, oflag: ::c_int, ...) -> ::c_int;
@@ -769,2 +778,3 @@
     )]
+    #[cfg_attr(gnu_time64_abi, link_name = "creat64")]
     pub fn creat(path: *const c_char, mode: mode_t) -> ::c_int;
@@ -774,2 +784,3 @@
     )]
+    #[cfg_attr(gnu_time64_abi, link_name = "__fcntl_time64")]
     pub fn fcntl(fd: ::c_int, cmd: ::c_int, ...) -> ::c_int;
@@ -796,2 +807,3 @@
     )]
+    #[cfg_attr(gnu_time64_abi, link_name = "readdir64")]
     pub fn readdir(dirp: *mut ::DIR) -> *mut ::dirent;
@@ -834,2 +846,3 @@
     )]
+    #[cfg_attr(gnu_time64_abi, link_name = "__fstatat64_time64")]
     pub fn fstatat(
@@ -912,2 +925,3 @@
     pub fn link(src: *const c_char, dst: *const c_char) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "lseek64")]
     pub fn lseek(fd: ::c_int, offset: off_t, whence: ::c_int) -> off_t;
@@ -940,2 +954,3 @@
     #[cfg_attr(target_os = "netbsd", link_name = "__nanosleep50")]
+    #[cfg_attr(gnu_time64_abi, link_name = "__nanosleep64")]
     pub fn nanosleep(rqtp: *const timespec, rmtp: *mut timespec) -> ::c_int;
@@ -970,2 +985,3 @@
     )]
+    #[cfg_attr(gnu_time64_abi, link_name = "pread64")]
     pub fn pread(fd: ::c_int, buf: *mut ::c_void, count: ::size_t, offset: off_t) -> ::ssize_t;
@@ -975,2 +991,3 @@
     )]
+    #[cfg_attr(gnu_time64_abi, link_name = "pwrite64")]
     pub fn pwrite(fd: ::c_int, buf: *const ::c_void, count: ::size_t, offset: off_t) -> ::ssize_t;
@@ -979,2 +996,3 @@
     #[cfg_attr(target_os = "netbsd", link_name = "__utime50")]
+    #[cfg_attr(gnu_time64_abi, link_name = "__utime64")]
     pub fn utime(file: *const c_char, buf: *const utimbuf) -> ::c_int;
@@ -1001,2 +1019,3 @@
     )]
+    #[cfg_attr(gnu_time64_abi, link_name = "mmap64")]
     pub fn mmap(
@@ -1027,2 +1046,3 @@
     )]
+    #[cfg_attr(gnu_time64_abi, link_name = "__lstat64_time64")]
     pub fn lstat(path: *const c_char, buf: *mut stat) -> ::c_int;
@@ -1049,3 +1069,5 @@
 
+    #[cfg_attr(gnu_time64_abi, link_name = "truncate64")]
     pub fn truncate(path: *const c_char, length: off_t) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "ftruncate64")]
     pub fn ftruncate(fd: ::c_int, length: off_t) -> ::c_int;
@@ -1055,2 +1077,3 @@
     #[cfg_attr(target_os = "netbsd", link_name = "__getrusage50")]
+    #[cfg_attr(gnu_time64_abi, link_name = "__getrusage64")]
     pub fn getrusage(resource: ::c_int, usage: *mut rusage) -> ::c_int;
@@ -1132,2 +1155,3 @@
     )]
+    #[cfg_attr(gnu_time64_abi, link_name = "__pthread_cond_timedwait64")]
     pub fn pthread_cond_timedwait(
@@ -1188,2 +1212,3 @@
     #[cfg_attr(target_os = "espidf", link_name = "lwip_getsockopt")]
+    #[cfg_attr(gnu_time64_abi, link_name = "__getsockopt64")]
     pub fn getsockopt(
@@ -1198,2 +1223,3 @@
     #[cfg_attr(target_os = "netbsd", link_name = "__utimes50")]
+    #[cfg_attr(gnu_time64_abi, link_name = "__utimes64")]
     pub fn utimes(filename: *const ::c_char, times: *const ::timeval) -> ::c_int;
@@ -1256,2 +1282,3 @@
     // FIXME: for `time_t`
+    #[cfg_attr(gnu_time64_abi, link_name = "__gmtime64_r")]
     pub fn gmtime_r(time_p: *const time_t, result: *mut tm) -> *mut tm;
@@ -1260,2 +1287,3 @@
     // FIXME: for `time_t`
+    #[cfg_attr(gnu_time64_abi, link_name = "__localtime64_r")]
     pub fn localtime_r(time_p: *const time_t, result: *mut tm) -> *mut tm;
@@ -1268,2 +1296,3 @@
     // FIXME: for `time_t`
+    #[cfg_attr(gnu_time64_abi, link_name = "__mktime64")]
     pub fn mktime(tm: *mut tm) -> time_t;
@@ -1272,2 +1301,3 @@
     // FIXME: for `time_t`
+    #[cfg_attr(gnu_time64_abi, link_name = "__time64")]
     pub fn time(time: *mut time_t) -> time_t;
@@ -1276,2 +1306,3 @@
     // FIXME: for `time_t`
+    #[cfg_attr(gnu_time64_abi, link_name = "__gmtime64")]
     pub fn gmtime(time_p: *const time_t) -> *mut tm;
@@ -1280,2 +1311,3 @@
     // FIXME: for `time_t`
+    #[cfg_attr(gnu_time64_abi, link_name = "__localtime64")]
     pub fn localtime(time_p: *const time_t) -> *mut tm;
@@ -1284,2 +1316,3 @@
     // FIXME: for `time_t`
+    #[cfg_attr(gnu_time64_abi, link_name = "__difftime64")]
     pub fn difftime(time1: time_t, time0: time_t) -> ::c_double;
@@ -1288,2 +1321,3 @@
     // FIXME: for `time_t`
+    #[cfg_attr(gnu_time64_abi, link_name = "__timegm64")]
     pub fn timegm(tm: *mut ::tm) -> time_t;
@@ -1342,2 +1376,3 @@
     #[cfg_attr(target_os = "netbsd", link_name = "__select50")]
+    #[cfg_attr(gnu_time64_abi, link_name = "__select64")]
     pub fn select(
@@ -1360,3 +1395,5 @@
     pub fn sem_post(sem: *mut sem_t) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "statvfs64")]
     pub fn statvfs(path: *const c_char, buf: *mut statvfs) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "fstatvfs64")]
     pub fn fstatvfs(fd: ::c_int, buf: *mut statvfs) -> ::c_int;
@@ -1383,3 +1420,5 @@
 
+    #[cfg_attr(gnu_time64_abi, link_name = "fseeko64")]
     pub fn fseeko(stream: *mut ::FILE, offset: ::off_t, whence: ::c_int) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "ftello64")]
     pub fn ftello(stream: *mut ::FILE) -> ::off_t;
@@ -1400,2 +1439,3 @@
     pub fn tcsendbreak(fd: ::c_int, duration: ::c_int) -> ::c_int;
+    #[cfg_attr(gnu_time64_abi, link_name = "mkstemp64")]
     pub fn mkstemp(template: *mut ::c_char) -> ::c_int;
@@ -1424,2 +1464,3 @@
 
+    #[cfg_attr(gnu_time64_abi, link_name = "lockf64")]
     pub fn lockf(fd: ::c_int, cmd: ::c_int, len: ::off_t) -> ::c_int;
@@ -1434,2 +1475,3 @@
         extern "C" {
+            #[cfg_attr(gnu_time64_abi, link_name = "__adjtime64")]
             pub fn adjtime(delta: *const timeval, olddelta: *mut timeval) -> ::c_int;
@@ -1478,2 +1520,3 @@
                           mode: ::mode_t) -> ::c_int;
+            #[cfg_attr(gnu_time64_abi, link_name = "openat64")]
             pub fn openat(dirfd: ::c_int, pathname: *const ::c_char,
@@ -1501,2 +1544,3 @@
             /// https://www.unix.com/man-page/opensolaris/3LIB/libc/
+            #[cfg_attr(gnu_time64_abi, link_name = "readdir64_r")]
             pub fn readdir_r(dirp: *mut ::DIR, entry: *mut ::dirent,
@@ -1554,2 +1598,3 @@
             #[cfg_attr(target_os = "netbsd", link_name = "__pselect50")]
+            #[cfg_attr(gnu_time64_abi, link_name = "__pselect64")]
             pub fn pselect(
